lib/OpusVL/FB11/Auth/Brain.pm

```diff
@@ -22,18 +55,127 @@ has short_name => (
+
+=cut
+
 # TODO retest legacy parameters against someone who has them
 sub hats {
     (
         qw/auth parameters/,
         fb11authdb => {
             class => 'auth'
-        }
+        },
+        'dbicdh::consumer' => {
+            class => '+OpusVL::FB11::Hat::dbicdh::consumer::is_brain',
+            constructor => {
+                priority => 0,
+            }
+        },
+        'dbicdh::manager',
     )
 }
 
 sub provided_services {
-    qw/auth fb11authdb/
+    qw/auth fb11authdb dbicdh::manager/
+}
+
+# TODO: I don't like this being done automatically but we have to make sure DBs
+# are deployed before other inits run
+sub hive_init {
+    my $self = shift;
+    my $hive = shift;
+
+    $hive->service('dbicdh::manager')->deploy_and_upgrade($hive);
+
+    # XXX TEMPORARY
+    # This is not where this should live but this is sort of the right place.
+    # The correct behaviour would be a UI that does this on request, which we
+    # can direct to if we have no auth data yet.
+    $self->_setup_fb11admin($hive);
+}
+
+sub _setup_fb11admin {
+    my $self = shift;
+    my $hive = shift;
+
+    # If we don't have an app, just don't do it. This only matters if there's an
+    # app to provide auth to (although why are you here if there's no app?)
+    my $app = try { $hive->service('fb11::app') } catch { undef }
+        or return;
+
+    my $feature_list = $app->auth_feature_list;
+
+    # XXX Because this is temporary I'm not bothering to see if something else
+    # is providing auth. The ability to change auth provider is way further
+    # ahead than the ability to set up FB11 on first run.
+
+    # Find any user. If there is none, create basic fb11admin user.
+    my $Users = $self->schema->resultset('User');
+    my $user = $Users->first;
+
+    return if $user;
+
+    $user = $Users->create({
+        username => 'fb11admin',
+        email    => 'fb11admin@localhost',
+        name     => 'Administrator',
+        password => 'fb11password'
+    });
+
+    warn "**** Created fb11admin / fb11password! Change the password!";
+
+    for my $section (keys %$feature_list) {
+        my $roles = $feature_list->{$section};
+
+        for my $role (keys %$roles) {
+            $self->schema->resultset('Aclfeature')->find_or_create({
+                feature => $section . '/' . $role
+            });
+        }
+    }
+    my $admin_role = $self->schema->resultset('Role')->find_or_create({
+        role => 'Admin'
+    });
+
+    $self->schema->txn_do(sub {
+        for my $feature ( $self->schema->resultset('Aclfeature')->all ) {
+            $admin_role->add_to_aclfeatures($feature)
+                unless $admin_role->aclfeatures->find($feature->id);
+        }
+    });
+
+    $user->add_to_roles($admin_role);
+
+    # This is bullshit. FB11 caches all the stuff but I have no idea how and the
+    # ACL is dumb as shit. We can remove most of this code anyway when we get a
+    # proper ACL so fuck it.
```

Comment:
could this be a bit embarassing if leaked?

Reply:
Perhaps ... Maybe it's time for me to stop being automatically embarrassed by everything in this repo now that good stuff is happening